applications/minotari_app_grpc/src/authentication/basic_auth.rs

```diff
@@ -77,17 +92,40 @@ impl BasicAuthCredentials {
         Ok(credentials)
     }
 
+    // This function provides a constant time comparison of the given username with the registered username.
+    fn constant_time_compare_username(&self, username: &str) -> bool {
+        let a_bytes = self.user_name.as_bytes();
+        let b_bytes = username.as_bytes();
+
+        let valid_username = !(b_bytes.is_empty() || b_bytes.len() > MAX_USERNAME_LEN);
+        let b_bytes_len_clipped = min(b_bytes.len(), MAX_USERNAME_LEN);
+
+        // Comparison bytes for both usernames are initialized to a large array of equal bytes.
+        let mut a_compare_bytes = [7u8; MAX_USERNAME_LEN];
+        let mut b_compare_bytes = a_compare_bytes;
+
+        // Add the comparison bytes for the actual username to the respective comparison byte arrays.
+        a_compare_bytes[0..a_bytes.len()].clone_from_slice(a_bytes);
+        b_compare_bytes[0..b_bytes_len_clipped].clone_from_slice(&b_bytes[..b_bytes_len_clipped]);
+
+        (a_compare_bytes.ct_eq(&b_compare_bytes).unwrap_u8() != 0) && valid_username
+    }
+
     pub fn validate(&self, username: &str, password: &[u8]) -> Result<(), BasicAuthError> {
-        if self.user_name.as_bytes() != username.as_bytes() {
-            return Err(BasicAuthError::InvalidUsername);
-        }
+        let valid_username = self.constant_time_compare_username(username);
+
         // These bytes can leak if the password is not utf-8, but since argon encoding is utf-8 the given
         // password must be incorrect if conversion to utf-8 fails.
         let bytes = self.password.reveal().to_vec();
         let str_password = Zeroizing::new(String::from_utf8(bytes)?);
         let header_password = PasswordHash::parse(&str_password, Encoding::B64)?;
-        Argon2::default().verify_password(password, &header_password)?;
-        Ok(())
+        let valid_password = Argon2::default().verify_password(password, &header_password).is_ok();
+
+        if valid_username && valid_password {
```

Comment:
I worry that the use of boolean logic here could open the door to the compiler being too clever for its own good.

You could use the result of `verify_password` to set a `Choice`, have `constant_time_compare_username` also return a `Choice`, and then use `BitAnd` logic on those instead. This will keep the logic behind the `subtle` optimization barriers.

Reply:
I also just noticed that `subtle` provides an [idiomatic way](https://docs.rs/subtle/latest/subtle/struct.Choice.html#impl-From%3CChoice%3E-for-bool) to get a `bool` from a `Choice` that would be useful for the final return value here.